Generalize asynchronous events

ShockedPlot7560 opened the pull request:
## Introduction
This PR would be more of a foretaste for #6015 in that it provides a means of generalizing asynchronous event calling.

An asynchronous event in this context is one whose action can be deferred by plugins specifying promises to be resolved.  
The integrity of event priorities is guaranteed by calling them one by one. Before moving on to the next priority, we wait until all the promises of the current one have been fulfilled.  
Only the state of the server and player context can't be guaranteed. However, the state of the event is unaffected, whether it is called asynchronous or synchronous.

## Behaviour

When a listener of an asynchronous event returns a `Promise`, the event is marked as asynchronous and placed at the end of the call stack for its priority. 
When a callAsync is called, the call flow is as follows:
```
1. for each priority:
   2. call async handlers which allows concurrent execution
   3. for each async handlers which doesn't allows concurrent execution:
       1. wait for previous promises
       2. clear all the promise
       3. call the handler and add the return promise into the waiting list
   4. wait for all promises
2. the promise returned by callAsync is resolved with their instance
```

In this way, each listener can add only one promise.
If, on the other hand, the handler is sensitive to changes in the event, it is preferable to add the @exclusiveCall tag, which specifies that it must be the only one being called.

## Event transformation

If we wanted to implement the asynchronous chat event, this would look like this: 
```php
class PlayerChatAsyncEvent extends AsyncEvent implements Cancellable{
	use CancellableTrait;
... Logic here
}
```
When calling: 
```php
$ev = new PlayerChatAsyncEvent($this, $messagePart, $this->server->getBroadcastChannelSubscribers(Server::BROADCAST_CHANNEL_USERS), new StandardChatFormatter());
$ev->callAsync()->onCompletion(function(PlayerChatAsyncEvent $ev) : void{
	if(!$ev->isCancelled()){
		$this->server->broadcastMessage($ev->getFormatter()->format($ev->getPlayer()->getDisplayName(), $ev->getMessage()), $ev->getRecipients());
	}
}, function() {
	// NOOP
});
```

## Changes
### API changes
- Add `AsyncEvent` class
- Add `Promise::all()`

### Behavioural changes


## Backwards compatibility
Not for the moment

## Follow-up
n/a

## Tests
```php
class Main extends PluginBase implements Listener{
	protected function onEnable() : void{
		$this->getServer()->getPluginManager()->registerEvents($this, $this);
	}

	public function onPlayerChatAsyncConc1(PlayerChatEvent $event) : Promise {
		return $this->timeout("Player " . $event->getPlayer()->getName() . " chat async concurrent 1");
	}

	public function onPlayerChatAsyncConc2(PlayerChatEvent $event) : Promise {
		return $this->timeout("Player " . $event->getPlayer()->getName() . " chat async concurrent 2");
	}

	/**
	 * @exclusiveCall
	 */
	public function onPlayerChatAsyncNoConc(PlayerChatEvent $event) : Promise {
		return $this->timeout("Player " . $event->getPlayer()->getName() . " chat async exclusiveCall");
	}

	private function timeout(string $message) : Promise {
		$promise = new PromiseResolver();

		$this->getScheduler()->scheduleDelayedTask(new ClosureTask(function() use ($message, $promise) {
			$this->getLogger()->debug("Resolving promise for message " . $message);
			$promise->resolve(null);
		}), 5 * 20);

		return $promise->getPromise();
	}
}
```

ShockedPlot7560 committed:
temporaly add Promise::all
code came from #6015 and will be subject to probable changes

ShockedPlot7560 committed:
introduce AsyncEvent and ::callAsync()
An asynchronous event is one that allows the addition of promises to be resolved before being completed.
This implementation integrates priority levels, allowing you to wait for all promises added to one level before moving on to the next.
This is made possible by separating the event call logic into several functions, which can then be integrated into AsyncEventTrait::callAsync()

SOF3 commented on src/promise/Promise.php:
```diff
@@ -23,10 +23,11 @@
- * @phpstan-template TValue
+ * @phpstan-template-covariant TValue
  */
 final class Promise{
```
So are we now determined to turn `Promise` into a general-purpose async util class?

SOF3 commented on src/promise/Promise.php:
```diff
@@ -23,10 +23,11 @@
 
 /**
- * @phpstan-template TValue
+ * @phpstan-template-covariant TValue
```
are you sure a covariant `TValue` really makes sense? it seems that `TValue` would be contravariant in `onCompletion`.

ShockedPlot7560 replied:
Without, all analystic tools will raised some error when we pass null on mixed attended type

dktapps replied:
I'd also prefer if it could be avoided. It's confusing and I feel like it's probably not necessary now that we can do `Promise<*>` in newer versions of PHPStan.

SOF3 commented on src/promise/Promise.php:
```diff
@@ -52,4 +53,54 @@ public function onCompletion(\Closure $onSuccess, \Closure $onFailure) : void{
+		$continue = true;
+
+		foreach($promises as $key => $promise){
+			$values[$key] = null;
```
Why do we need to initialize them as null?

SOF3 commented on src/promise/Promise.php:
```diff
@@ -52,4 +53,54 @@ public function onCompletion(\Closure $onSuccess, \Closure $onFailure) : void{
+	 * `$promises` indexed by the respective Promises' array keys.
+	 *
+	 * @template TPromiseValue
+	 * @phpstan-param Promise<TPromiseValue>[] $promises
```
What about preserving the key here? `array<K, Promise<T>>` -> `Promise<array<K, T>>`

SOF3 commented on src/event/AsyncEventTrait.php:
```diff
@@ -0,0 +1,84 @@
+use function array_shift;
+use function count;
+
+trait AsyncEventTrait {
```
Consider extracting the contents of this trait into a class and just proxy the methods to the class. Adding too many implementation details in a trait is dangerous since traits are just copied code.

SOF3 commented on src/event/AsyncEventTrait.php:
```diff
@@ -0,0 +1,84 @@
+		$this->promises = $promises;
+	}
+
+	public function addPromise(Promise $promise) : void{
```
Is it an expected use case that a single handler adds multiple promises?

ShockedPlot7560 replied:
Yes, this will be handled as if multiple handlers were adding promises. 

Added promises are only processed once all handlers of the priority have been called, so no difference will be made on this side.

SOF3 replied:
What if user calls `addPromise` on the event again after the initial dispatch? Your `Promise::all()` call won't include the new promises, and we cannot prevent them from doing so.

ShockedPlot7560 replied:
We should throw an exception if `addPromise` is called in an asynchronous context, to avoid this confusion.

SOF3 commented on src/event/AsyncEventTrait.php:
```diff
@@ -0,0 +1,84 @@
+			$array = $promiseSet->toArray();
+			$promiseSet->clear();
+
+			return Promise::all($array);
```
So multiple handlers of the same priority can execute concurrently? Would this easily lead to developers "forgetting" to check if another handler cancelled the event during execution, or multiple handlers concurrently overwriting each other's data? (could happen if there are suspension points between setter calls)

SOF3 commented on src/event/Event.php:
```diff
@@ -47,15 +47,30 @@ final public function getEventName() : string{
 		$timings->startTiming();
 
-		$handlers = HandlerListManager::global()->getHandlersFor(static::class);
+		$handlers = $handlers ?? HandlerListManager::global()->getHandlersFor(static::class);
```
Why not just pass this directly?

dktapps commented on src/event/AsyncEventTrait.php:
```diff
@@ -0,0 +1,84 @@
+		$this->promises->add($promise);
+	}
+
+	final public static function callAsync(AsyncEvent&Event $event, ObjectSet $promiseSet) : Promise{
```
why is this static? we're in a trait

ShockedPlot7560 committed:
events: asynchandler is defined by their return type and event type
If the event is async and the handler return Promise, it will be handle as an async event.
However, if the event is async and the handler return smth different than Promise, it will be handle synchronously.
An async handler can specify if it wishes to be called with no concurrent handlers with the tag @noConcurrentCall

ShockedPlot7560 committed:
oops, remove test code

ShockedPlot7560 committed:
fix style

ShockedPlot7560 committed:
undo Promise covariant + improve array types

ShockedPlot7560 committed:
fix PHPstan

ShockedPlot7560 commented:
## New behaviour

Above all, the logic of synchronous events remains the same, even for asynchronous events, as long as they are called synchronously.

When a listener of an asynchronous event returns a `Promise`, the event is marked as asynchronous and placed at the end of the call stack for its priority. 
When a callAsync is called, the call flow is as follows:
```
1. for each priority:
   1. call sync handlers (return type is different than Promise).
   2. call async handlers which allows concurrent execution
   3. for each async handlers which doesn't allows concurrent execution:
       1. wait for previous promises
       2. clear all the promise
       3. call the handler and add the return promise into the waiting list
   4. wait for all promises
2. the promise returned by callAsync is resolved
```

In this way, each listener can add only one promise.
If, on the other hand, the handler is sensitive to changes in the event, it is preferable to add the @noConcurrentCall tag, which specifies that it must be the only one being called.

ShockedPlot7560 committed:
phpstan: populate baseline

SOF3 commented on src/event/AsyncEvent.php:
```diff
@@ -34,20 +33,11 @@
 	 * @phpstan-return Promise<null>
 	 */
-	public static function callAsync(AsyncEvent&Event $event, ObjectSet $promiseSet) : Promise;
+	public function callAsync() : Promise;
```
Does this mean the same event class can be called either synchronously or asynchronously? What happens if PlayerChatEvent is dispatched with `call()` in another plugin?

ShockedPlot7560 replied:
Currently, both are possible.
I was wondering whether an exception should be thrown when someone tries to call an async event in sync.  
As a result, if an async event is called in sync, all the handlers are called without taking into account the return types, i.e. without waiting for the promises of the others, etc...

ShockedPlot7560 replied:
It is no longer possible to call an asynchronous event synchronously.

SOF3 commented on src/event/AsyncEventTrait.php:
```diff
@@ -24,61 +24,17 @@
 trait AsyncEventTrait {
-	/** @phpstan-var ObjectSet<Promise<null>> */
-	private ObjectSet $promises;
+	private AsyncEventDelegate $delegate;
```
Beware GC performance impact caused by circular references. Is it even necessary to store this field in the event object?

JavierLeon9966 replied:
That can be solved if `AsyncEventDelegate` stores a `WeakReference` of the class.

SOF3 replied:
The point is that the circular reference was never necessary to begin with. Why store an object you don't use?

JavierLeon9966 replied:
Right. That class can just be `AsyncEventTrait`.

JavierLeon9966 replied:
Right. That class can just be `AsyncEventTrait`.

ShockedPlot7560 replied:
> The point is that the circular reference was never necessary to begin with. Why store an object you don't use?

I agree, the class property is not necessary

SOF3 commented on src/event/HandlerList.php:
```diff
@@ -128,11 +130,26 @@ public function getListenerList() : array{
 		}
+		foreach($asyncListenersByPriority as $priority => $asyncListeners){
+			usort($asyncListeners, static function(RegisteredAsyncListener $a, RegisteredAsyncListener $b) : int{
+				if($a->canBeCallConcurrently()){
```
Grammar issue? Either `canCall` or `canBeCalled`, not `canBeCall`

SOF3 commented on src/event/HandlerList.php:
```diff
@@ -128,11 +130,26 @@ public function getListenerList() : array{
+					return $b->canBeCallConcurrently() ? 0 : -1;
+				}else{
+					return $b->canBeCallConcurrently() ? -1 : 0;
+				}
```
Can we simplify this function into `$a->canBeCallConcurrently() <=> $b->canBeCallConcurrently()`?

ShockedPlot7560 replied:
Can be simplify into `($a->canBeCallConcurrently() <=> $b->canBeCallConcurrently()) * -1` yes

JavierLeon9966 replied:
That's just `$b->canBeCallConcurrently() <=> $a->canBeCallConcurrently()`.

SOF3 commented on src/event/HandlerList.php:
```diff
@@ -128,11 +130,26 @@ public function getListenerList() : array{
+		foreach($asyncListenersByPriority as $priority => $asyncListeners){
+			usort($asyncListeners, static function(RegisteredAsyncListener $a, RegisteredAsyncListener $b) : int{
+				if($a->canBeCallConcurrently()){
+					return $b->canBeCallConcurrently() ? 0 : -1;
```
Is it arbitrary that exclusive handlers are called first and concurrent handlers are called together last? Would like to see the justification in comments.

ShockedPlot7560 replied:
I have no justification for the fact that these are the exclusive handlers after the concurrent handlers because I haven't been able to make up my mind? The only performance requirement is that the concurrent handlers be grouped together and that the exclusive handlers be called either before or after the concurrent ones.

SOF3 commented on src/event/ListenerMethodTags.php:
```diff
@@ -31,4 +31,5 @@ final class ListenerMethodTags{
 	public const HANDLE_CANCELLED = "handleCancelled";
 	public const NOT_HANDLER = "notHandler";
 	public const PRIORITY = "priority";
+	public const NO_CONCURRENT_CALL = "noConcurrentCall";
```
Concurrent call is actually a less intuitive behavior. I would rather have concurrent handlers opt-in.

btw it is generally an antipattern to use negative names in any programming symbol. Prefer using `exclusiveCall` if we stick with opt-in excusive.

SOF3 commented on src/plugin/PluginManager.php:
```diff
@@ -626,8 +629,29 @@ public function registerEvents(Listener $listener, Plugin $plugin) : void{
+					));
+				}
+				switch(strtolower($tags[ListenerMethodTags::NO_CONCURRENT_CALL])){
+					case "true":
```
since we have multiple of these, what about extracting to a function?

ShockedPlot7560 replied:
This is not directly related to this PR, it should be done externally.

SOF3 commented on src/plugin/PluginManager.php:
```diff
@@ -653,8 +677,25 @@ public function registerEvent(string $event, \Closure $handler, int $priority, P
 		$timings = Timings::getEventHandlerTimings($event, $handlerName, $plugin->getDescription()->getFullName());
 
-		$registeredListener = new RegisteredListener($handler, $priority, $plugin, $handleCancelled, $timings);
+		if(is_subclass_of($event, AsyncEvent::class) && $this->canHandleAsyncEvent($handler)){
```
`registerEvent` is a low-level API that does not perform any reflection checks on `$handler`. It doesn't look right to start introducing magical behavior to `registerEvent`; that's the job of `registerEvents`.

SOF3 commented on src/plugin/PluginManager.php:
```diff
@@ -653,8 +677,25 @@ public function registerEvent(string $event, \Closure $handler, int $priority, P
+		$reflection = new \ReflectionFunction($handler);
+		$return = $reflection->getReturnType();
+
+		return $return instanceof \ReflectionNamedType && $return->getName() === Promise::class;
```
Any thoughts on making async handlers opt-in in general, and just check the return type to double-confirm? i.e. either handler return type is `Promise`, it does not return a `Promise` object and it does not declare `@async`, or it declares `@async` and return type is `Promise`; if it has a return type hint of `Promise` or the function invocation returns a `Promise`, the server should crash.

ShockedPlot7560 replied:
As async events are now completely detached from synchronous events, there is no longer any need for double confirmation, as each event handler will behave in the same way.

SOF3 commented on src/plugin/PluginManager.php:
```diff
@@ -626,8 +629,29 @@ public function registerEvents(Listener $listener, Plugin $plugin) : void{
+						break;
+					default:
+						throw new PluginException("Event handler " . Utils::getNiceClosureName($handlerClosure) . "() declares invalid @" . ListenerMethodTags::NO_CONCURRENT_CALL . " value \"" . $tags[ListenerMethodTags::NO_CONCURRENT_CALL] . "\"");
+				}
```
Not related to this PR, but both the API and this handler would be much nicer if we use attributes instead:

```php
#[\pocketmine\event\AsyncHandler]
public function handler(XxxEvent $event) : Promise { ... }

#[\pocketmine\event\AsyncHandler(concurrent: true)]
public function handler(XxxEvent $event) : Promise { ... }
```

SOF3 reviewed (commented):
tldr AsyncEventDelegate, will look again when PR is ready

JavierLeon9966 commented on src/player/Player.php:
```diff
@@ -1517,6 +1517,7 @@ public function chat(string $message) : bool{
 					if(!$ev->isCancelled()){
 						$this->server->broadcastMessage($ev->getFormatter()->format($ev->getPlayer()->getDisplayName(), $ev->getMessage()), $ev->getRecipients());
 					}
+
```
```suggestion
```

dktapps commented on src/event/Event.php:
```diff
@@ -47,6 +47,21 @@ final public function getEventName() : string{
+				$registration->callEvent($this);
+			}
+		});
+	}
```
This code is very performance sensitive. I don't like introducing extra function calls and closures into here.

dktapps committed:
Update src/player/Player.php

Co-authored-by: Javier León <58715544+JavierLeon9966@users.noreply.github.com>

dktapps reviewed (commented):
It seems to me like a lot of complexity went into making async events and their handlers use the same frameworks that sync events do. The net effect of this is that async events are callable either synchronously or asynchronously, which seems to me like a design flaw.

ShockedPlot7560 commented:
> It seems to me like a lot of complexity went into making async events and their handlers use the same frameworks that sync events do. The net effect of this is that async events are callable either synchronously or asynchronously, which seems to me like a design flaw.

We can prohibit the synchronous call of a specified asynchronous event, but exiting the event will require a BC break every time we want to implement it in an event. As I said [here](https://github.com/pmmp/PocketMine-MP/pull/6092#discussion_r1367932362) synchronous and asynchronous calling is a subject for discussion, however, it's still necessary to keep synchronous handlers in the middle of synchronous ones, as not everyone needs to return a promise every time.

SOF3 commented:
I think what we need here is an adapter at the registerEvents level to convert a void function into a promise closure. imo we are fine with breaking BC for registerEvent on PlayerChatEvent as long as we don't break registerEvents, and it is probably fine to make PlayerChatEvent no longer extend Event.

ShockedPlot7560 committed:
merge remote tracking

ShockedPlot7560 committed:
cannot call async event in sync context + remove Event dependency for AsyncEventDelegate

ShockedPlot7560 committed:
clarifying the exception message

ShockedPlot7560 committed:
correction of various problems

ShockedPlot7560 committed:
move the asynchronous registration of handlers to a dedicated PluginManager function

ShockedPlot7560 committed:
fix PHPStan

ShockedPlot7560 committed:
inconsistency correction

ShockedPlot7560 committed:
follow up of #6110

ShockedPlot7560 committed:
handlerListe: reduce code complexity

JavierLeon9966 commented on src/event/HandlerList.php:
```diff
@@ -129,24 +128,23 @@ public function getListenerList() : array{
-			});
-			$listenersByPriority[$priority] = array_merge($listenersByPriority[$priority] ?? [], $asyncListeners);
-		}
+		$listenersByPriority = array_merge_recursive($listeners, $asyncListeners, $exclusiveAsyncListeners);
```
Aren't exclusive async listeners supposed to be run before normal async listeners?

ShockedPlot7560 replied:
That's the debate going on in my neurons. I don't have any pros or cons, really.

dktapps replied:
The defining feature of exclusive listeners is that no other listener can be invoked before the exclusive listener's promise is resolved. The ordering isn't really significant.

dktapps replied:
Looks like this function isn't actually used for async events anyway.

dktapps commented on src/event/Event.php:
```diff
@@ -51,6 +51,9 @@ public function call() : void{
 		}
+		if($this instanceof AsyncEvent){
+			throw new \InvalidArgumentException("Cannot call async event synchronously, use callAsync() instead");
+		}
```
Again, we're still sharing architecture here. This function should simply not exist for an async event.

dktapps commented:
`Promise::all()` should be PR'd separately, as it's useful by itself and would reduce PR diff noise

ShockedPlot7560 committed:
remove using of Event API

ShockedPlot7560 committed:
Merge remote-tracking branch 'upstream/minor-next' into feat/async-events

ShockedPlot7560 committed:
merge remote tracking upstream

ShockedPlot7560 committed:
resolve AsyncEvent with self instance

ShockedPlot7560 committed:
fix PHPstan

ShockedPlot7560 commented:
## New behavior (again)

As async event is BC breaking by default, I detached the async event from the synchronous ones. Now, an async event should be handle with an async handler : must return a Promise. It can have a tag to specify if he need to be an exclusive handler or not. 
Exclusive handlers will be called one by one, so no other handler can modify the event during the execution. Otherwise, concurrent handler will be call in parallel, priority by priority.

ShockedPlot7560 committed:
Merge remote-tracking branch 'origin/minor-next' into feat/async-events

dktapps commented on src/event/HandlerListManager.php:
```diff
@@ -112,7 +113,7 @@ public function getListFor(string $event) : HandlerList{
 
 	/**
-	 * @phpstan-param class-string<covariant Event> $event
+	 * @phpstan-param class-string<Event|AsyncEvent> $event
```
Shouldn't this be templated like getListFor()?

dktapps commented on src/event/RegisteredAsyncListener.php:
```diff
@@ -0,0 +1,67 @@
+use pocketmine\promise\Promise;
+use pocketmine\timings\TimingsHandler;
+
+class RegisteredAsyncListener extends RegisteredListener{
```
I don't think this should inherit from RegisteredListener. There's no commonality aside from the shared properties.

dktapps commented:
So, this is pretty much how I envisioned it should be. I think it lost me somewhere because this whole concept is just a massive headache.

To summarize what I think I'm seeing:

# Promise rejection
If an async handler at any priority rejects the promise it returned, no handlers from subsequent priorities will be called.
If an exclusive async handler rejects, no subsequent exclusive handlers will be called, and no handlers from subsequent priorities will be called either.

This is a "hard cancellation" with no ability to uncancel. Unsure if this is desirable, or how we might change this in the future if the need arose. Please discuss.

# Concurrent vs exclusive handlers
There are 2 types of async handler: concurrent (all concurrent priorities with the same priority will be run together and then their promises collectively awaited), and exclusive (each promise must resolve before the next exclusive handler is called).

**Use case explanations & documentation needed for this.**

# Separated stack for sync vs async events
Async events and sync events are fully separated. Sync events cannot have async handlers and vice versa. I personally requested this.

This avoids impact to existing events, as they otherwise would ALL need to be post-processed asynchronously (which would require massive refactors & break any plugin that calls events of its own), or the functionality would need to be enabled for select events (which causes weird problems of its own).

This also avoids breaking existing sync handler code, which weren't designed for the possibility that 2 events of the same type might be in an incomplete state at the same time. This means they aren't built to expect that, for example `LOWEST` priority handlers could be called 2 times with a certain event before `HIGHEST` priority handlers were called at all.

**However**, I think it makes sense to allow async events to have sync handlers. We just need to be aware that switching an event from sync to async or vice versa is always a BC break in one form or another.

# Caller of events
Callers of events can simply put an `onCompletion` callback pair onto the promise returned by `call()`. This is nice and simple.

# Event handlers
Async event handlers must return `Promise`. This is different from sync events, whose return types are not enforced, but which are assumed to return `void`.

Otherwise, async handlers look basically the same as sync handlers.

dktapps commented on src/plugin/PluginManager.php:
```diff
@@ -665,4 +693,44 @@ public function registerEvent(string $event, \Closure $handler, int $priority, P
+		$reflection = new \ReflectionFunction($handler);
+		$return = $reflection->getReturnType();
+
+		return $return instanceof \ReflectionNamedType && $return->getName() === Promise::class;
```
This probably needs to check for `allowsNull()`. Not sure what should be done about that.

dktapps replied:
In fact I don't think we actually need to check this at all. Sync handlers on async events should be fine, it's only the inverse that's a problem.

dktapps committed:
Merge branch 'minor-next' of github.com:pmmp/PocketMine-MP into feat/async-events

dktapps committed:
Merge branch 'minor-next' into feat/async-events

dktapps committed:
Update PHPStan baseline

dktapps committed:
Fully separate hierarchies for sync & async events
there's no way to combine these without causing type compatibility issues for one side or the other.
we might be able to use traits to reduce duplication, but the separation here seems to be necessary.

dktapps committed:
HandlerListManager: added getter

dktapps committed:
optimization

dktapps committed:
Fix PHPStan error

dktapps committed:
Add integration tests
most of these are failing - needs to be investigated

dktapps committed:
Fixed bug in concurrency integration test

dktapps committed:
AsyncEvent: make the code easier to make sense of

dktapps committed:
Fix doc comments

dktapps committed:
CS

dktapps committed:
Fixed test code
the test still doesn't pass, but at least it's actually testing the problem now...

dktapps committed:
ah hello my old friend, impossible-generics.neon
propagating generics all the way through the likes of HandlerList etc is more trouble than it's worth.

dktapps committed:
Handler inheritance is now working
this code should also perform somewhat better

dktapps committed:
cleanup

dktapps committed:
Reduce code duplication

dktapps committed:
PluginManager: ensure that handler candidates of async events with wrong return types don't attempt to register as sync events
this will cause other, more confusing errors to be thrown.

to be honest, I'm not sure if enforcing the return type here is even necessary (or desirable).

dktapps committed:
Split AsyncHandlerListManager
this allows further code deduplication at the expense of needing 2 calls to unregister all handlers

dktapps committed:
Reduce number of classes

dktapps committed:
...

dktapps committed:
Remove unused class

dktapps committed:
CONTRIBUTING: changing an event from sync to async or vice versa is a BC break

dktapps committed:
Merge branch 'minor-next' into feat/async-events

SOF3 commented on src/event/AsyncEvent.php:
```diff
@@ -0,0 +1,126 @@
+abstract class AsyncEvent{
+	/** @var array<class-string<AsyncEvent>, int> $delegatesCallDepth */
+	private static array $delegatesCallDepth = [];
+	private const MAX_EVENT_CALL_DEPTH = 50;
```
What if I dispatch 50 AsyncPlayerLoginEvents within 1 second when server starts, and the handlers take more than 1 second to resolve? The server crashes immediately?

dktapps replied:
I discussed this with Plot on Discord yesterday. I don't think even he knows how this code was supposed to work.

ShockedPlot7560 replied:
Probably TODO now, but we probably should track where all handlers stuck and throw exception if too many handlers stuck at one handler for too long? 
As said on discord, this can produce some crash if a server handler like 100 connections at a time and the login sequence is transformed into async process.

SOF3 commented on src/event/AsyncEvent.php:
```diff
@@ -0,0 +1,126 @@
+			}else{
+				if(count($awaitPromises) > 0){
+					//wait for concurrent promises to complete
+					break;
```
Should this be `continue` instead of `break` to process other concurrent handlers at the same priority? Or do we sort handlers by `(priority, isExclusive)`?

SOF3 replied:
Oh I see it's sorted indeed.

dktapps replied:
Functionally it's not actually important anyway. The sorting only affects the number of times Promise::all() will be called.

SOF3 commented on src/event/AsyncEvent.php:
```diff
@@ -0,0 +1,126 @@
+	private function processRemainingHandlers(array $handlers, \Closure $resolve, \Closure $reject) : void{
+		$currentPriority = null;
+		$awaitPromises = [];
+		foreach($handlers as $k => $handler){
```
This function is a bit verbose. Can we do this better, e.g. treating exclusive handlres as their own batches and letting handler list return a list of batches to execute?

dktapps replied:
Linear cache bypasses the handler list entirely in most cases for performance reasons, same as for sync events. If we start doing what you suggest it will just require even more code.

SOF3 commented on src/event/AsyncHandlerListManager.php:
```diff
@@ -0,0 +1,61 @@
+	private static function sortSamePriorityHandlers(array $listeners) : array{
+		uasort($listeners, function(AsyncRegisteredListener $left, AsyncRegisteredListener $right) : int{
+			//While the system can handle these in any order, it's better for latency if concurrent handlers
+			//are processed together. It doesn't matter whether they are processed before or after exclusive handlers.
+			if($right->canBeCalledConcurrently()){
+				return $left->canBeCalledConcurrently() ? 0 : 1;
+			}
+			return -1;
```
But you end up defining an order here right? Ends up `$left > $right` if `$left` is exclusive and `$right` is concurrent.

Also why do we return `-1` when both handlers are exclusive? Shouldn't it return `0` instead?

```suggestion
			//are processed together. Exclusive handlers are sorted before concurrent handlers.
			$leftConcurrent = $left->canBeCalledConcurrently();
			$rightConcurrent = $right->canBeCalledConcurrently();
			if($leftConcurrent !== $rightConcurrent){
				return $leftConcurrent <=> $rightConcurrent;
			}
			return 0;
```

dktapps replied:
Yes, the order is intended to group concurrent handlers together, whether at the start or the end.

Good catch on the sameness. I knew there was something I didn't like about this code but I couldn't put my finger on it.

dktapps replied:
I don't like specifying that exclusive handlers should come first without reasoning. There is no reason for them to come first. What we actually want is to aggregate concurrent handlers for Promise::all() to be efficiently used.

dktapps committed:
Simplify handler sorting

ShockedPlot7560 commented on src/event/AsyncEvent.php:
```diff
@@ -0,0 +1,126 @@
+			$globalResolver = new PromiseResolver();
+
+			$handlers = AsyncHandlerListManager::global()->getHandlersFor(static::class);
+			if(count($handlers) > 0){
```
This check can be avoided because it will be checked in the processRemainingHandlers function and solved directly?

dktapps replied:
It's a micro optimization to avoid unnecessary allocations

ShockedPlot7560 commented:
As I understand, now, we will accept async listeners to return null or Promise ?

SenseiTarzan commented:
I think the event system in asynchronous will never work properly because it was not designed for it. At the time, we couldn't do asynchronous operations, and because of that, the entire pmmp codebase would need to be reworked to integrate the asynchronous part into pmmp's event loop for proper handling.
Because we can perform synchronous operations within asynchronous code, but not the reverse, especially since managing cancellations would not be possible in a very clean way.
You have to do it before PM6

dktapps commented:
> As I understand, now, we will accept async listeners to return null or Promise ?

Yep, they'll allow the next handler in the queue to be called immediately if returning null

dktapps committed:
Merge branch 'minor-next' into feat/async-events

dktapps committed:
Convert integration tests to unit tests
this required mocking to get around #6524. Longer term we should make improvements to avoid the need for mocking here.

dktapps committed:
CS

dktapps committed:
we don't need a fake server instance outside of setUp() in these tests

dktapps commented:
I think the main thing still on the TODO list is to buffer incomplete events. The existing mechanism using stack depth won't cut it. We need a way to track which handlers didn't fulfil their promises so that the correct plugin can be properly blamed when an event fails to complete.

dktapps committed:
Merge branch 'minor-next' into feat/async-events

dktapps committed:
fix PHPStan error

dktapps committed:
github web editor don't fuck up indentation, challenge impossible

ShockedPlot7560 committed:
Merge remote-tracking branch 'upstream/minor-next' into feat/async-events

ShockedPlot7560 commented:
To be more precise with the depth limiter. We could probably reset it after each concurrent management of priority handlers. There's less chance of a plugin making a recursion call with different priority handlers.
I'm not saying we'll cover all cases because that's impossible, but we can limit errors and that's necessary to protect the integrity of the server at least.

ShockedPlot7560 commented:
I said nothing, mind is blowing

ShockedPlot7560 committed:
Implement handlers stuck detection system

ShockedPlot7560 commented:
A question remains about the global asynchronous events timing. Should we keep it as it will result in an inconsistent result as it profiles when we call the event until an asynchronous handler is encountered? The handler call is always profiled until a promise is returned, perhaps, like asyncTask, we need to profile the handler time in general? 
Open to opinion.

ShockedPlot7560 commented:
I modified the system to detect handlers blocked at a point. 
Now, each time a handler is called, a personal counter is increased by 1 until its execution is complete. This ensures that there are not too many handlers blocked at a point, which is why the value has been doubled. 
It won't stop the server crashing when 1000 players connect at the same time when a plugin has added a long task behind it. 
Perhaps an extra function in the AsyncEvent so that you can give your personal limit? Perhaps a security issue? But for me, it's OK as it is here.

dktapps commented on src/event/AsyncEvent.php:
```diff
@@ -0,0 +1,134 @@
+			}
+
+			$currentPriority = $priority;
+			$handlerId = spl_object_id($handler) << 3 | $priority;
```
Why is the priority included here? This seems risky and will silently break if arbitrary priority numbers are added in the future.

We shouldn't need priority to correctly blame a handler for not fulfilling event calls, should we?

ShockedPlot7560 replied:
The handler registry ensures that there are no two identical instances of the same handler in the same priority. However, between different priorities, we may encounter the same handler. Should we consider handlers to be identical even if their priority is different?

dktapps replied:
That's impossible, their `RegisteredListener` instances would be different even if the same `Closure` was reused

dktapps committed:
Merge branch 'minor-next' into feat/async-events

ShockedPlot7560 committed:
Compare handlers only by their IDs

ShockedPlot7560 committed:
add concurrent calls test

ShockedPlot7560 committed:
phpstan :/

ShockedPlot7560 committed:
Fix PHPstan & CS

ShockedPlot7560 committed:
thanks impossible-generics.neon

dktapps committed:
Fix phpdoc types

ShockedPlot7560 committed:
Add more unit tests for Listener

ShockedPlot7560 committed:
remove pointless test

dktapps committed:
Merge branch 'minor-next' into feat/async-events

dktapps committed:
...

dktapps committed:
Merge branch 'feat/async-events' of github.com:pmmp/PocketMine-MP into feat/async-events

dktapps committed:
shut up bruv
